modules/ublog/src/main/UblogBestOf.scala

```diff
@@ -0,0 +1,100 @@
+
+import scala.util.Try
+
+import reactivemongo.api.bson.BSONNull
+
+import scalalib.paginator.{ AdapterLike, Paginator }
+
+import lila.db.dsl.{ *, given }
+import lila.memo.CacheApi
+
+object UblogBestOf:
+
+  private val ublogOrigin = YearMonth.of(2021, 9)
+  private def nbMonthsBackward =
+    ublogOrigin.until(currentYearMonth, ChronoUnit.MONTHS).toInt
+
+  private def currentYearMonth = YearMonth.now(ZoneOffset.UTC)
+  def allYears                 = (ublogOrigin.getYear to currentYearMonth.getYear + 1).toList
+
+  def selector(month: YearMonth) =
+    val (start, end) = boundsOfMonth(month)
+    // to hit topic prod index
+    $doc("topics".$ne(UblogTopic.offTopic), "lived.at".$gt(start).$lt(end))
+
+  private def boundsOfMonth(month: YearMonth): (Instant, Instant) =
+    val start = month.atDay(1).atStartOfDay()
+    val end   = month.atEndOfMonth().atTime(LocalTime.MAX)
+    (start.toInstant(ZoneOffset.UTC), end.toInstant(ZoneOffset.UTC))
+
+  def readYear(year: Int): Option[Year] =
+    (ublogOrigin.getYear <= year && year <= currentYearMonth.getYear).so(Try(Year.of(year)).toOption)
+
+  def readYearMonth(year: Int, month: Int): Option[YearMonth] =
+    Try(YearMonth.of(year, month)).toOption.filter: ym =>
+      // writing it as negative allow bounds to be included
+      !(ym.isBefore(ublogOrigin) || ym.isAfter(currentYearMonth))
+
+  private def monthsBack(n: Int): YearMonth =
+    currentYearMonth.minusMonths(n)
+
+  // from `now` go back to `offset` months and from that point gives all `length` precedecing months
+  def slice(offset: Int, length: Int): Seq[(YearMonth, Int)] =
+    val from = currentYearMonth.minusMonths(offset)
+    (0 to length).map(x => from.minusMonths(x.toInt)).zip((0 to length))
+
+  case class WithPosts(yearMonth: YearMonth, posts: List[UblogPost.PreviewPost])
+
+final class UblogBestOf(colls: UblogColls, ublogApi: UblogApi, cacheApi: CacheApi)(using Executor):
+
+  import UblogBsonHandlers.{ *, given }
+
+  def paginatorQuery(offset: Int, length: Int): Fu[List[UblogBestOf.WithPosts]] =
+    colls.post
+      .aggregateList(length, _.sec): framework =>
+        import framework.*
+        Facet(
+          UblogBestOf
+            .slice(offset = offset, length = length)
+            .map: (month, i) =>
+              s"$i" -> (List(
+                Match($doc("live" -> true) ++ UblogBestOf.selector(month))
+              ) ++ UblogRank.Sorting.ByTimelessRank.sortingQuery(colls.post, framework) ++ List(
+                Limit(4)
+              ) ++ ublogApi.removeUnlistedOrClosed(colls.post, framework))
+        ) -> List(
+          Project($doc("all" -> $doc("$objectToArray" -> "$$ROOT"))),
+          UnwindField("all"),
+          ReplaceRootField("all"),
+          Project(
+            $doc(
+              "v"          -> true,
+              "monthsBack" -> $doc("$toInt" -> "$k")
+            )
+          ),
+          Sort(Ascending("monthsBack"))
+        )
```

Comment:
that is one BRUTAL aggregation query. https://gist.github.com/ornicar/01aff9112aca96fc3d1dca77aecdf602

Reply:
That's a neat DoS target. We need to cache it.